C630/qcom-gps/gpsd-master/man/gpsd.8
'\" t
.\"     Title: gpsd
.\"    Author: [see the "AUTHORS" section]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 9 Aug 2004
.\"    Manual: GPSD Documentation
.\"    Source: The GPSD Project
.\"  Language: English
.\"
.TH "GPSD" "8" "9 Aug 2004" "The GPSD Project" "GPSD Documentation"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
gpsd \- interface daemon for GPS receivers
.SH "SYNOPSIS"
.HP \w'\fBgpsd\fR\ 'u
\fBgpsd\fR [\-b] [\-D\ \fIdebuglevel\fR] [\-F\ \fIcontrol\-socket\fR] [\-f\ \fIframing\fR] [\-G] [\-h] [\-l] [\-n] [\-N] [\-P\ \fIpidfile\fR] [\-r] [\-S\ \fIlistener\-port\fR] [\-s\ \fIspeed\fR] [\-V] [[\fIsource\-name\fR]...]
.SH "QUICK START"
.PP
If you have a GPS attached on the lowest\-numbered USB port of a Linux system, and want to read reports from it on TCP/IP port 2947, it will normally suffice to do this:
.sp
.if n \{\
.RS 4
.\}
.nf
gpsd /dev/ttyUSB0
.fi
.if n \{\
.RE
.\}
.PP
For the lowest\-numbered serial port:
.sp
.if n \{\
.RS 4
.\}
.nf
gpsd /dev/ttyS0
.fi
.if n \{\
.RE
.\}
.PP
Change the device number as appropriate if you need to use a different port\&. Command\-line flags enable verbose logging, a control port, and other optional extras but should not be needed for basic operation; the one exception, on very badly designed hardware, might be
\fB\-b\fR
(which see)\&.
.PP
On Linux systems supporting udev,
gpsd
is normally started automatically when a USB plugin event fires (if it is not already running) and is handed the name of the newly active device\&. In that case no invocation is required at all\&.
.PP
For your initial tests set your GPS hardware to speak NMEA, as
gpsd
is guaranteed to be able to process that\&. If your GPS has a native or binary mode with better performance that
gpsd
knows how to speak,
gpsd
will autoconfigure that mode\&.
.PP
You can verify correct operation by first starting
gpsd
and then
xgps, the X windows test client\&.
.PP
If you have problems, the GPSD project maintains a FAQ to assist troubleshooting\&.
.SH "DESCRIPTION"
.PP
gpsd
is a monitor daemon that collects information from GPSes, differential\-GPS radios, or AIS receivers attached to the host machine\&. Each GPS, DGPS radio, or AIS receiver is expected to be direct\-connected to the host via a USB or RS232C serial device\&. The serial device may be specified to
gpsd
at startup, or it may be set via a command shipped down a local control socket (e\&.g\&. by a USB hotplug script)\&. Given a GPS device by either means,
gpsd
discovers the correct port speed and protocol for it\&.
.PP
gpsd
should be able to query any GPS that speaks either the standard textual NMEA 0183 protocol, or the (differing) extended NMEA dialects used by MKT\-3301, iTrax, Motorola OnCore, Sony CXD2951, and Ashtech/Thales devices\&. It can also interpret the binary protocols used by EverMore, Garmin, Navcom, Rockwell/Zodiac, SiRF, Trimble, and u\-blox ANTARIS devices\&. Under Linux it can read NMEA2000 packets through the kernel CAN socket\&. It can read heading and attitude information from the Oceanserver 5000 or TNT Revolution digital compasses\&.
.PP
The GPS reporting formats supported by your instance of
gpsd
may differ depending on how it was compiled; general\-purpose versions support many, but it can be built with protocol subsets down to a singleton for use in constrained environments\&. For a list of the GPS protocols supported by your instance, see the output of
\fBgpsd \-l\fR
.PP
gpsd
effectively hides the differences among the GPS types it supports\&. It also knows about and uses commands that tune these GPSes for lower latency\&. By using
gpsd
as an intermediary, applications avoid contention for serial devices\&.
.PP
gpsd
can use differential\-GPS corrections from a DGPS radio or over the net, from a ground station running a DGPSIP server or a Ntrip broadcaster that reports RTCM\-104 data; this will shrink position errors by roughly a factor of four\&. When
gpsd
opens a serial device emitting RTCM\-104, it automatically recognizes this and uses the device as a correction source for all connected GPSes that accept RTCM corrections (this is dependent on the type of the GPS; not all GPSes have the firmware capability to accept RTCM correction packets)\&. See
the section called \(lqACCURACY\(rq
and
the section called \(lqFILES\(rq
for discussion\&.
.PP
Client applications will communicate with
gpsd
via a TCP/IP port, 2947 by default)\&. Both IPv4 and IPv6 connections are supported and a client may connect via either\&.
.PP
The program accepts the following options:
.PP
\-b
.RS 4
Broken\-device\-safety mode, otherwise known as read\-only mode\&. A few bluetooth and USB receivers lock up or become totally inaccessible when probed or reconfigured; see the hardware compatibility list on the GPSD project website for details\&. This switch prevents gpsd from writing to a receiver\&. This means that
gpsd
cannot configure the receiver for optimal performance, but it also means that
gpsd
cannot break the receiver\&. A better solution would be for Bluetooth to not be so fragile\&. A platform independent method to identify serial\-over\-Bluetooth devices would also be nice\&.
.RE
.PP
\-D
.RS 4
Set debug level\&. At debug levels 2 and above,
gpsd
reports incoming sentence and actions to standard error if
gpsd
is in the foreground (\-N) or to syslog if in the background\&.
.RE
.PP
\-F
.RS 4
Create a control socket for device addition and removal commands\&. You must specify a valid pathname on your local filesystem; this will be created as a Unix\-domain socket to which you can write commands that edit the daemon\*(Aqs internal device list\&.
.RE
.PP
\-f
.RS 4
Fix the framing to the GNSS device\&. The framing parameter is of the form: [78][ENO][012]\&. Most GNSS are 8N1\&. Some Trimble default to 8O1\&. The default is to search for the correct framing\&.
.RE
.PP
\-G
.RS 4
This flag causes
gpsd
to listen on all addresses (INADDR_ANY) rather than just the loop back (INADDR_LOOPBACK) address\&. For the sake of privacy and security, TPV information is now private to the local machine until the user makes an effort to expose this to the world\&.
.RE
.PP
\-h
.RS 4
Display help message and terminate\&.
.RE
.PP
\-l
.RS 4
List all drivers compiled into this
gpsd
instance\&. The letters to the left of each driver name are the
gpsd
control commands supported by that driver\&.
.RE
.PP
\-n
.RS 4
Don\*(Aqt wait for a client to connect before polling whatever GPS is associated with it\&. Some RS232 GPSes wait in a standby mode (drawing less power) when the host machine is not asserting DTR, and some cellphone and handheld embedded GPSes have similar behaviors\&. Accordingly, waiting for a watch request to open the device may save battery power\&. (This capability is rare in consumer\-grade devices)\&. You should use this option if you plan to use gpsd to provide reference clock information to ntpd through a memory\-shared segment\&.
.RE
.PP
\-N
.RS 4
Don\*(Aqt daemonize; run in foreground\&. This switch is mainly useful for debugging\&.
.RE
.PP
\-r
.RS 4
Use GPS time even with no current fix\&. Some GPS\*(Aqs have battery powered Real Time Clocks (RTC\*(Aqs) built in, making them a valid time source even before a fix is acquired\&. This can be useful on a Raspberry Pi, or other device that has no battery powered RTC, and thus has no valid time at startup\&.
.RE
.PP
\-P
.RS 4
Specify the name and path to record the daemon\*(Aqs process ID\&.
.RE
.PP
\-S
.RS 4
Set TCP/IP port on which to listen for GPSD clients (default is 2947)\&.
.RE
.PP
\-s
.RS 4
Fix the speed to the GNSS device\&. The default is to autobaud\&.
.RE
.PP
\-V
.RS 4
Dump version and exit\&.
.RE
.PP
Arguments are interpreted as the names of data sources\&. Normally, a data source is the device pathname of a local device from which the daemon may expect GPS data\&. But there are three other special source types recognized, for a total of four:
.PP
Local serial or USB device
.RS 4
A normal Unix device name of a serial or USB device to which a sensor is attached\&. Example:
/dev/ttyUSB0\&.
.RE
.PP
Local PPS device
.RS 4
A normal Unix device name of a PPS device to which a PPS source is attached\&. The device name must start with "/dev/pps" and a local serial or USB GPS device must also be available\&. Example:
/dev/pps0\&.
.RE
.PP
TCP feed
.RS 4
A URI with the prefix "tcp://", followed by a hostname, a colon, and a port number\&. The daemon will open a socket to the indicated address and port and read data packets from it, which will be interpreted as though they had been issued by a serial device\&. Example:
tcp://data\&.aishub\&.net:4006\&.
.RE
.PP
UDP feed
.RS 4
A URI with the prefix "udp://", followed by a hostname, a colon, and a port number\&. The daemon will open a socket listening for UDP datagrams arriving on the indicated address and port, which will be interpreted as though they had been issued by a serial device\&. Example:
udp://127\&.0\&.0\&.1:5000\&.
.RE
.PP
Ntrip caster
.RS 4
A URI with the prefix "ntrip://" followed by the name of an Ntrip caster (Ntrip is a protocol for broadcasting differential\-GPS fixes over the net)\&. For Ntrip services that require authentication, a prefix of the form "username:password@" can be added before the name of the Ntrip broadcaster\&. For Ntrip service, you must specify which stream to use; the stream is given in the form "/streamname"\&. An example DGPSIP URI could be "dgpsip://dgpsip\&.example\&.com" and a Ntrip URI could be "ntrip://foo:bar@ntrip\&.example\&.com:80/example\-stream"\&. Corrections from the caster will be send to each attached GPS with the capability to accept them\&.
.RE
.PP
DGPSIP server
.RS 4
A URI with the prefix "dgpsip://" followed by a hostname, a colon, and an optional colon\-separated port number (defaulting to 2101)\&. The daemon will handshake with the DGPSIP server and read RTCM2 correction data from it\&. Corrections from the server will be set to each attached GPS with the capability to accept them\&. Example:
dgpsip://dgps\&.wsrcc\&.com:2101\&.
.RE
.PP
Remote gpsd feed
.RS 4
A URI with the prefix "gpsd://", followed by a hostname and optionally a colon and a port number (if the port is absent the default
gpsd
port will be used)\&. The daemon will open a socket to the indicated address and port and emulate a
gpsd
client, collecting JSON reports from the remote
gpsd
instance that will be passed to local clients\&.
.RE
.PP
NMEA2000 CAN data
.RS 4
A URI with the prefix "nmea2000://", followed by a CAN devicename\&. Only Linux socket CAN interfaces are supported\&. The interface must be configured to receive CAN messages before
gpsd
can be started\&. If there is more then one unit on the CAN bus that provides GPS data,
gpsd
chooses the unit from which a GPS message is first seen\&. Example:
nmea2000://can0\&.
.RE
.PP
(The "ais:://" source type supported in some older versions of the daemon has been retired in favor of the more general "tcp://"\&.)
.PP
Additionally, two serial device names have a side effect:
.PP
/dev/ttyAMA0
.RS 4
The UART device on a Raspberry Pi\&. Has the side effect of opening /dev/pps0 for RFC2783 1PPS data\&.
.RE
.PP
/dev/gpsd0
.RS 4

C630/qcom-gps/gpsd-master/gpsd.php
<?php

# DO NOT HAND_HACK! THIS FILE IS GENERATED

# Copyright (c) 2006,2010 Chris Kuethe <chris.kuethe@gmail.com>
#
# This file is Copyright (c) 2009-2019 The GPSD project
# SPDX-License-Identifier: BSD-2-clause

# Changed to Google Maps API v3, requires no API key, and shorter code
# Sanjeev Gupta <ghane0@gmail.com> 2013-01-05

global $head, $blurb, $title, $showmap, $autorefresh, $footer, $gmap_key;
global $server, $advertise, $port, $open, $swap_ew, $testmode;
$testmode = 1; # leave this set to 1

# Public script parameters:
#   host: host name or address where GPSd runs. Default: from config file
#   port: port of GPSd. Default: from config file
#   op=view: show just the skyview image instead of the whole HTML page
#     sz=small: used with op=view, display a small (240x240px) skyview
#   op=json: respond with the GPSd POLL JSON structure
#     jsonp=prefix: used with op=json, wrap the POLL JSON in parentheses
#                   and prepend prefix

# If you're running PHP with the Suhosin patch (like the Debian PHP5 package),
# it may be necessary to increase the value of the
# suhosin.get.max_value_length parameter to 2048. The imgdata parameter used
# for displaying the skyview is longer than the default 512 allowed by Suhosin.
# Debian has the config file at /etc/php5/conf.d/suhosin.ini.

# this script shouldn't take more than a few seconds to run
set_time_limit(3);
ini_set('max_execution_time', 3);

if (!file_exists("gpsd_config.inc"))
	write_config();

require_once("gpsd_config.inc");

# sample data
$resp = <<<EOF
{"class":"POLL","time":"2010-04-05T21:27:54.84Z","active":1,
 "tpv":[{"class":"TPV","tag":"MID41","device":"/dev/ttyUSB0",
           "time":1270517264.240,"ept":0.005,"lat":40.035093060,
           "lon":-75.519748733,"alt":31.1,"track":99.4319,
           "speed":0.123,"mode":3}],
 "sky":[{"class":"SKY","tag":"MID41","device":"/dev/ttyUSB0",
              "time":"2010-04-05T21:27:44.84Z","hdop":9.20,"vdop":12.1,
              "satellites":[{"PRN":16,"el":55,"az":42,"ss":36,"used":true},
                            {"PRN":19,"el":25,"az":177,"ss":0,"used":false},
                            {"PRN":7,"el":13,"az":295,"ss":0,"used":false},
                            {"PRN":6,"el":56,"az":135,"ss":32,"used":true},
                            {"PRN":13,"el":47,"az":304,"ss":0,"used":false},
                            {"PRN":23,"el":66,"az":259,"ss":40,"used":true},
                            {"PRN":20,"el":7,"az":226,"ss":0,"used":false},
                            {"PRN":3,"el":52,"az":163,"ss":32,"used":true},
                            {"PRN":31,"el":16,"az":102,"ss":0,"used":false}
                           ]
             }
            ]
}
EOF;



# if we're passing in a query, let's unpack and use it
$op = isset($_GET['op']) ? $_GET['op'] : '';
if (isset($_GET['imgdata']) && $op == 'view'){
	$resp = base64_decode($_GET['imgdata']);
	if ($resp){
		gen_image($resp);
		exit(0);
	}
} else {
	if (isset($_GET['host']))
		if (!preg_match('/[^a-zA-Z0-9\.-]/', $_GET['host']))
			$server = $_GET['host'];

	if (isset($_GET['port']))
		if (!preg_match('/\D/', $_GET['port']) && ($port>0) && ($port<65536))
			$port = $_GET['port'];

	if ($testmode){
		$sock = @fsockopen($server, $port, $errno, $errstr, 2);
		@fwrite($sock, "?WATCH={\"enable\":true}\n");
		usleep(1000);
		@fwrite($sock, "?POLL;\n");
		usleep(1000);
		for($tries = 0; $tries < 10; $tries++){
			$resp = @fread($sock, 2000); # SKY can be pretty big
			if (preg_match('/{"class":"POLL".+}/i', $resp, $m)){
				$resp = $m[0];
				break;
			}
		}
		@fclose($sock);
		if (!$resp)
			$resp = '{"class":"ERROR","message":"no response from GPS daemon"}';
	}
}

if ($op == 'view')
	gen_image($resp);
else if ($op == 'json')
	write_json($resp);
else
	write_html($resp);

exit(0);

###########################################################################
# Function to decide if a PRN is a true GPS bird or SBAS, GBAS, etc.
# Sanjeev Gupta <ghane0@gmail.com> 20150408
# Please refer to gps.h lines ~~ 95 , for a central definition
function isGPS($PRN) {
	if ($PRN <= 32) return TRUE ; 			# Navstar GPS
	if ($PRN >= 64 && $PRN <= 96) return TRUE ; 	# GLONASS
	if ($PRN >= 159 ) return TRUE ; 		# BeiDou ?
	return FALSE ;					# SBAS, GBAS, unknown
	}

function colorsetup($im){
	$C['white']	= imageColorAllocate($im, 255, 255, 255);
	$C['ltgray']	= imageColorAllocate($im, 191, 191, 191);
	$C['mdgray']	= imageColorAllocate($im, 127, 127, 127);
	$C['dkgray']	= imageColorAllocate($im, 63, 63, 63);
	$C['black']	= imageColorAllocate($im, 0, 0, 0);
	$C['red']	= imageColorAllocate($im, 255, 0, 0);
	$C['brightgreen'] = imageColorAllocate($im, 0, 255, 0);
	$C['darkgreen']	= imageColorAllocate($im, 0, 192, 0);
	$C['blue']	= imageColorAllocate($im, 0, 0, 255);
	$C['cyan']	= imageColorAllocate($im, 0, 255, 255);
	$C['magenta']	= imageColorAllocate($im, 255, 0, 255);
	$C['yellow']	= imageColorAllocate($im, 255, 255, 0);
	$C['orange']	= imageColorAllocate($im, 255, 128, 0);

	return $C;
}

function legend($im, $sz, $C){
	$r = 30;
	$fn = 5;
	$x = $sz - (4*$r+7) - 2;
	$y = $sz - $r - 3;

	imageFilledRectangle($im, $x, $y, $x + 4*$r + 7, $y + $r +1, $C['dkgray']);
	imageRectangle($im, $x+0*$r+1, $y+1, $x + 1*$r + 0, $y + $r, $C['red']);
	imageRectangle($im, $x+1*$r+2, $y+1, $x + 2*$r + 2, $y + $r, $C['yellow']);
	imageRectangle($im, $x+2*$r+4, $y+1, $x + 3*$r + 4, $y + $r, $C['darkgreen']);
	imageRectangle($im, $x+4*$r+6, $y+1, $x + 3*$r + 6, $y + $r, $C['brightgreen']);
	imageString($im, $fn, $x+3+0*$r, $y+$r/3, "<30", $C['red']);
	imageString($im, $fn, $x+5+1*$r, $y+$r/3, "30+", $C['yellow']);
	imageString($im, $fn, $x+7+2*$r, $y+$r/3, "35+", $C['darkgreen']);
	imageString($im, $fn, $x+9+3*$r, $y+$r/3, "40+", $C['brightgreen']);
}

function radial($angle, $sz){
	#turn into radians
	$angle = deg2rad($angle);

	# determine length of radius
	$r = $sz * 0.5 * 0.95;

	# and convert length/azimuth to cartesian
	$x0 = sprintf("%d", (($sz * 0.5) - ($r * cos($angle))));
	$y0 = sprintf("%d", (($sz * 0.5) - ($r * sin($angle))));
	$x1 = sprintf("%d", (($sz * 0.5) + ($r * cos($angle))));
	$y1 = sprintf("%d", (($sz * 0.5) + ($r * sin($angle))));

	return array($x0, $y0, $x1, $y1);
}

function azel2xy($az, $el, $sz){
	global $swap_ew;
	#rotate coords... 90deg E = 180deg trig
	$az += 270;

	#turn into radians
	$az = deg2rad($az);

	# determine length of radius
	$r = $sz * 0.5 * 0.95;
	$r -= ($r * ($el/90));

	# and convert length/azimuth to cartesian
	$x = sprintf("%d", (($sz * 0.5) + ($r * cos($az))));
	$y = sprintf("%d", (($sz * 0.5) + ($r * sin($az))));
	if ($swap_ew != 0)
		$x = $sz - $x;

	return array($x, $y);
}

function splot($im, $sz, $C, $e){
	if ((0 == $e['PRN']) || (0 == $e['az'] + $e['el']) ||
	    ($e['az'] < 0) || ($e['el'] < 0))
		return;

	$color = $C['brightgreen'];
	if ($e['ss'] < 40)
		$color = $C['darkgreen'];
	if ($e['ss'] < 35)
		$color = $C['yellow'];
	if ($e['ss'] < 30)
		$color = $C['red'];
	if ($e['el']<10)
		$color = $C['blue'];
	if ($e['ss'] < 10)
		$color = $C['black'];

	list($x, $y) = azel2xy($e['az'], $e['el'], $sz);

	$r = 12;
	if (isset($_GET['sz']) && ($_GET['sz'] == 'small'))
		$r = 8;

	imageString($im, 3, $x+4, $y+4, $e['PRN'], $C['black']);
        if ($e['used'] == true)
		if (isGPS($e['PRN']))
			imageFilledArc($im, $x, $y, $r, $r, 0, 360, $color, 0);
		else
			imageFilledDiamond($im, $x, $y, $r, $color);
	else
		if (isGPS($e['PRN']))
			imageArc($im, $x, $y, $r, $r, 0, 360, $color);
		else
			imageDiamond($im, $x, $y, $r, $color);
}


function imageDiamond($im, $x, $y, $r, $color){
	$t = $r/2;
	# this lunacy is because imagesetthickness doesn't seem to work
	$vx = array ( $x+$t, $y, $x, $y+$t, $x-$t, $y, $x, $y-$t );
	imagepolygon($im, $vx, 4, $color);
	$t--;
	$vx = array ( $x+$t, $y, $x, $y+$t, $x-$t, $y, $x, $y-$t );
	imagepolygon($im, $vx, 4, $color);
	$t--;
	$vx = array ( $x+$t, $y, $x, $y+$t, $x-$t, $y, $x, $y-$t );
	imagepolygon($im, $vx, 4, $color);
}

function imageFilledDiamond($im, $x, $y, $r, $color){
	$t = $r/2;
	while($t){
		$vx = array ( $x+$t, $y, $x, $y+$t, $x-$t, $y, $x, $y-$t );
		imagepolygon($im, $vx, 4, $color);
		$t -= 0.5;
	}
}

function elevation($im, $sz, $C, $a){
	$b = 90 - $a;
	$a = $sz * 0.95 * ($a/180);
	imageArc($im, $sz/2, $sz/2, $a*2, $a*2, 0, 360, $C['ltgray']);
	$x = $sz/2 - 16;
	$y = $sz/2 - $a;
	imageString($im, 2, $x, $y, $b, $C['ltgray']);
}

function skyview($im, $sz, $C){
	global $swap_ew;
	$a = 90; $a = $sz * 0.95 * ($a/180);
	imageFilledArc($im, $sz/2, $sz/2, $a*2, $a*2, 0, 360, $C['mdgray'], 0);
	imageArc($im, $sz/2, $sz/2, $a*2, $a*2, 0, 360, $C['black']);
	$x = $sz/2 - 16; $y = $sz/2 - $a;
	imageString($im, 2, $x, $y, "0", $C['ltgray']);

	$a = 85; $a = $sz * 0.95 * ($a/180);
	imageFilledArc($im, $sz/2, $sz/2, $a*2, $a*2, 0, 360, $C['white'], 0);
	imageArc($im, $sz/2, $sz/2, $a*2, $a*2, 0, 360, $C['ltgray']);
	imageString($im, 1, $sz/2 - 6, $sz+$a, '5', $C['black']);
	$x = $sz/2 - 16; $y = $sz/2 - $a;
	imageString($im, 2, $x, $y, "5", $C['ltgray']);

	for($i = 0; $i < 180; $i += 15){
		list($x0, $y0, $x1, $y1) = radial($i, $sz);
		imageLine($im, $x0, $y0, $x1, $y1, $C['ltgray']);
	}

	for($i = 15; $i < 90; $i += 15)
		elevation($im, $sz, $C, $i);

	$x = $sz/2 - 16; $y = $sz/2 - 8;
	/* imageString($im, 2, $x, $y, "90", $C['ltgray']); */

	imageString($im, 4, $sz/2 + 4, 2        , 'N', $C['black']);
	imageString($im, 4, $sz/2 + 4, $sz - 16 , 'S', $C['black']);
	if ($swap_ew != 0){
		imageString($im, 4, 4        , $sz/2 + 4, 'E', $C['black']);
		imageString($im, 4, $sz - 10 , $sz/2 + 4, 'W', $C['black']);
	} else {
		imageString($im, 4, 4        , $sz/2 + 4, 'W', $C['black']);
		imageString($im, 4, $sz - 10 , $sz/2 + 4, 'E', $C['black']);
	}
}

function gen_image($resp){
	$sz = 600;
	if (isset($_GET['sz']) && ($_GET['sz'] == 'small'))
		$sz = 240;

	$GPS = json_decode($resp, true);
	if ($GPS['class'] != "POLL"){
		die("json_decode error: $resp");
	}

	$im = imageCreate($sz, $sz);
	$C = colorsetup($im);
	skyview($im, $sz, $C);
	if ($sz > 240)
		legend($im, $sz, $C);

	for($i = 0; $i < count($GPS['sky'][0]['satellites']); $i++){
		splot($im, $sz, $C, $GPS['sky'][0]['satellites'][$i]);
	}

	header("Content-type: image/png");
	imagePNG($im);
	imageDestroy($im);
}

function dfix($x, $y, $z){
	if ($x < 0){
		$x = sprintf("%f %s", -1 * $x, $z);
	} else {
		$x = sprintf("%f %s", $x, $y);
	}
	return $x;
}

function write_html($resp){
	global $sock, $errstr, $errno, $server, $port, $head, $body, $open;
	global $blurb, $title, $autorefresh, $showmap, $gmap_key, $footer;
	global $testmode, $advertise;

	$GPS = json_decode($resp, true);
	if ($GPS['class'] != 'POLL'){
		die("json_decode error: $resp");
	}

	header("Content-type: text/html; charset=UTF-8");

	global $lat, $lon;
	$lat = (float)$GPS['tpv'][0]['lat'];
	$lon = (float)$GPS['tpv'][0]['lon'];
	$x = $server; $y = $port;
	$imgdata = base64_encode($resp);
	$server = $x; $port = $y;

	if ($autorefresh > 0)
		$autorefresh = "<meta http-equiv='Refresh' content='$autorefresh'/>";
	else
		$autorefresh = '';

	$map_head = $map_body = $map_code = '';
	if ($showmap == 1) {
		$map_head = gen_gmap_head();
		$map_body = 'onload="Load()" onunload="GUnload()"';
		$map_code = gen_map_code();
	} else if ($showmap == 2) {
		$map_head = gen_osm_head();
		$map_body = 'onload="Load()"';
		$map_code = gen_map_code();
	}
	$part1 = <<<EOF
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
    "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="https://www.w3.org/1999/xhtml">
<head>
{$head}
{$map_head}
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>
<meta http-equiv="Content-Language" content="en,en-us"/>
<title>{$title} - GPSD Test Station {$lat}, {$lon}</title>
{$autorefresh}
<style>
.warning {
    color: #FF0000;
}

.fixed {
    font-family: mono-space;
}

.caption {
    text-align: left;
    margin: 1ex 3em 1ex 3em; /* top right bottom left */
}

.administrivia {
    font-size: small;
    font-family: verdana, sans-serif;
}
</style>
</head>

<body {$body} {$map_body}>
<center>
<table border="0">
<tr><td align="justify">
{$blurb}
</td>
EOF;

	if (!strlen($advertise))
		$advertise = $server;

	if ($testmode && !$sock)
		$part2 = "";
	else
		$part2 = <<<EOF
<!-- ------------------------------------------------------------ -->

<td rowspan="4" align="center" valign="top">
<img src="?op=view&amp;imgdata={$imgdata}"
width="600" height="600"/>
<br clear="all"/>
<p class="caption">A filled circle means the satellite was used in
the last fix. Green-yellow-red colors indicate signal strength in dB,
green=most and red=least.  Diamonds indicate Augmentation satellites.</p>
{$map_code}</td>
</tr>
EOF;

	if (!$open)
		$part3 = '';
	else
		$part3 = <<<EOF
<!-- ------------------------------------------------------------ -->

<tr><td align="justify">To get real-time information, connect to
<span class="fixed">telnet://{$advertise}:{$port}/</span> and type "?POLL;"
or "?WATCH={"enable":true,"raw":true}".<br/>
Use a different server:<br/>
<form method=GET action="${_SERVER['SCRIPT_NAME']}">
<input name="host" value="{$advertise}">:
<input name="port" value="{$port}" size="5" maxlength="5">
<input type=submit value="Get Position"><input type=reset></form>
<br/>
</td>
</tr>
EOF;

	if ($testmode && !$sock)
		$part4 = "<tr><td class='warning'>The gpsd instance that this page monitors is not running.</td></tr>";
	else {
		$fix = $GPS['tpv'][0];
		$sky = $GPS['sky'][0];
		$sats = $sky['satellites'];

		$fixtype = array('Unknown' => 0, 'No Fix' => 1,'2D Fix' => 2, '3D Fix' => 3);
		$type = array_search($fix['mode'],$fixtype);

		$nsv = count($sats);
                $ts = $fix['time'];
                $sat = '';
                foreach($sats as $s)
                        $sat .= sprintf(
                                "\t<tr><td>%d</td><td>%d</td><td>%d</td><td>%d</td><td>%s</td></tr>\n",
                                $s['PRN'], $s['el'], $s['az'], $s['ss'], $s['used'] ? 'Y' : 'N'
                        );
		$part4 = <<<EOF
<!-- ------------------------------------------------------------ -->
<tr><td align=center valign=top>
    <table border=1>
        <tr><th colspan=2 align=center>Current Information</th></tr>
        <tr><td>Time (UTC)</td><td>{$ts}</td></tr>
        <tr><td>Latitude</td><td>{$fix['lat']}</td></tr>
        <tr><td>Longitude</td><td>{$fix['lon']}</td></tr>
        <tr><td>Altitude</td><td>{$fix['alt']}</td></tr>
        <tr><td>Fix Type</td><td>{$type}</td></tr>
        <tr><td>Satellites</td><td>{$nsv}</td></tr>
        <tr><td>HDOP</td><td>{$sky['hdop']}</td></tr>
        <tr><td>VDOP</td><td>{$sky['vdop']}</td></tr>
    </table>
    <br/>
    <table border=1>
        <tr><th colspan=5 align=center>Current Satellites</th></tr>
        <tr><th>PRN</th><th>Elevation</th><th>Azimuth</th><th>SS</th><th>Used</th></tr>
$sat    </table>
</td></tr>

<!-- raw response:
{$resp}
-->
EOF;
	}

	$part5 = <<<EOF

</table>
</center>
{$footer}
<hr/>
<p class="administrivia">This script is distributed by the
<a href="https://gpsd.io/">GPSD project</a>.</p>

</body>
</html>
EOF;

print $part1 . $part2 . $part3 . $part4 . $part5;

}

function write_json($resp){
	header('Content-Type: text/javascript');
	if (isset($_GET['jsonp']))
		print "{$_GET['jsonp']}({$resp})";
	else
		print $resp;
}

function write_config(){
	$f = fopen("gpsd_config.inc", "a");
	if (!$f)
		die("can't generate prototype config file. try running this script as root in DOCUMENT_ROOT");

	$buf = <<<EOB
<?PHP
\$title = 'My GPS Server';
\$server = 'localhost';
#\$advertise = 'localhost';
\$port = 2947;
\$autorefresh = 0; # number of seconds after which to refresh
\$showmap = 0; # set to 1 if you want to have a google map, set it to 2 if you want a map based on openstreetmap
\$gmap_key = 'GetYourOwnGoogleKey'; # your google API key goes here
\$swap_ew = 0; # set to 1 for upward facing view (nonstandard)
\$open = 0; # set to 1 to show the form to change the GPSd server

## You can read the header, footer and blurb from a file...
# \$head = file_get_contents('/path/to/header.inc');
# \$body = file_get_contents('/path/to/body.inc');
# \$footer = file_get_contents('/path/to/footer.hinc');
# \$blurb = file_get_contents('/path/to/blurb.inc');

## ... or you can just define them here
\$head = '';
\$body = '';
\$footer = '';
\$blurb = <<<EOT
This is a
<a href="https://gpsd.io/">gpsd</a>
server <blink><font color="red">located someplace</font></blink>.

The hardware is a
<blink><font color="red">hardware description and link</font></blink>.

This machine is maintained by
<a href="mailto:you@example.com">Your Name Goes Here</a>.<br/>
EOT;

?>

EOB;
	fwrite($f, $buf);
	fclose($f);
}

function gen_gmap_head() {
global $gmap_key;
return <<<EOT
<script src="//maps.googleapis.com/maps/api/js?sensor=false"/>

<script>
 <!--
    function Load() {
      var map = new google.maps.Map(
        document.getElementById('map'), {
          center: new google.maps.LatLng({$GLOBALS['lat']}, {$GLOBALS['lon']}),
          zoom: 13,
          mapTypeId: google.maps.MapTypeId.ROADMAP
      });

      var marker = new google.maps.Marker({
            position: new google.maps.LatLng({$GLOBALS['lat']}, {$GLOBALS['lon']}),
            map: map
      });

    }
    google.maps.event.addDomListener(window, 'load', initialize);

 -->
</script>
EOT;
}

function gen_osm_head() {
global $GPS;
return <<<EOT
<script src="http://openlayers.org/api/OpenLayers.js" />
<script>
    <!--
function Load() {
	document.getElementById("map").firstChild.data = "";
	var map = new OpenLayers.Map("map", {
		controls: [
			new OpenLayers.Control.Navigation(),
			new OpenLayers.Control.PanZoomBar(),
			new OpenLayers.Control.ScaleLine(),
			new OpenLayers.Control.LayerSwitcher()
		]
	});
	var layer = new OpenLayers.Layer.OSM("Open Street Map");
	map.addLayer(layer);

	var center = new OpenLayers.LonLat({$GLOBALS['lon']}, {$GLOBALS['lat']})
		.transform(new OpenLayers.Projection("EPSG:4326"), map.getProjectionObject());
	map.setCenter(center, 12);

	var markers = new OpenLayers.Layer.Markers("Markers");
	markers.addMarker(new OpenLayers.Marker(center));
	map.addLayer(markers);
}
    -->
    </script>
EOT;
}

function gen_map_code() {
return <<<EOT
<br/>
<div id="map" style="width: 550px; height: 400px; border:1px; border-style: solid;">
    Loading...
    <noscript>
        <span class='warning'>Sorry: you must enable javascript to view our maps.</span><br/>
    </noscript>
</div>

EOT;
}

?>
